geometry/src/polygon.rs

```diff
@@ -138,15 +139,69 @@ impl Polygon {
     }
 
     /// Returns the polygon's edges, i.e. the lines between vertices, as vectors.
-    pub fn edges(&self) -> Vec<Vector> {
+    pub fn edges(&self) -> impl Iterator<Item = Vector> + '_ {
         let vertices = self.vertices();
         let shifted_vertices = vertices.iter().cycle().skip(1).take(vertices.len());
         vertices
             .iter()
             .zip(shifted_vertices)
             .map(|(&first_vertex, &second_vertex)| second_vertex - first_vertex)
             .map(Vector::from)
-            .collect()
+    }
+
+    fn scalar_project_onto_unit_vector(&self, axis: Vector) -> (f64, f64) {
+        let projection: Vec<_> = self
+            .vertices()
+            .iter()
+            .cloned()
+            .map(Vector::from)
+            .map(|vector| vector.dot_product(axis))
+            .collect();
+        (
+            *projection
```

Comment:
Could you sort this and then take the first and last elements?

Reply:
calling min(n) and max(n) is a O(n) algorithm. Quicksort is O(n log n). O(n) is more efficient: https://www.wolframalpha.com/input/?i=f(x)+%3D+x+and+g(x)+%3D+x+*+log+x+for+x+%3D+0+to+x+%3D+100